soaktest/soaktest.go

```
@@ -0,0 +1,258 @@
+	for _, broker := range raw["brokers"].(map[string]interface{}) {
+		var parts string
+		var avg float64
+		var p99 float64
+		var nodeid int
+
+		topparsValue := broker.(map[string]interface{})["toppars"]
+		if topparsValue == nil ||
+			len(topparsValue.(map[string]interface{})) == 0 {
+			continue
+		}
+
+		for _, val := range topparsValue.(map[string]interface{}) {
+			if val.(map[string]interface{})["topic"] != topic {
+				continue
+			}
+			partition := val.(map[string]interface{})["partition"]
+			f, ok := partition.(float64)
+			if !ok {
+				ErrorLogger.Printf("Failed to convert partition "+
+					"to float64 %v\n", ok)
+				return ok
+			}
+			if len(parts) > 0 {
+				parts += "*"
+			}
+			parts += fmt.Sprintf("%d", int(f))
+		}
+
+		rttValue := broker.(map[string]interface{})["rtt"]
+		avg, ok := rttValue.(map[string]interface{})["avg"].(float64)
+		if !ok {
+			ErrorLogger.Printf("Failed to convert avg to float64 %v\n", ok)
+			return ok
+		}
+
+		p99, ok = rttValue.(map[string]interface{})["p99"].(float64)
+		if !ok {
+			ErrorLogger.Printf("Failed to convert p99 to float64 %v\n", ok)
+			return ok
+		}
+
+		nid, ok := broker.(map[string]interface{})["nodeid"].(float64)
+		if !ok {
+			ErrorLogger.Printf("Failed to convert nodeid to float64 %v\n", ok)
+			return ok
+		}
+		nodeid = int(nid)
+
+		tags := []string{fmt.Sprintf("broker:%d type: %s topic: %s parts: %s",
+			nodeid, raw["type"], topic, parts)}
+
+		DatadogGauge(BROKER_RTT_P99, p99/1000000.0, tags)
+		DatadogGauge(BROKER_RTT_AVG, avg/1000000.0, tags)
+	}
+	return true
+}
+
+// HandleStatsEvent converts Stats events to hashmap
+func HandleStatsEvent(e *kafka.Stats, topic string) error {
+	var raw map[string]interface{}
+	err := json.Unmarshal([]byte(e.String()), &raw)
+	if err != nil {
+		ErrorLogger.Printf("Json unmarshall error: %s\n", err)
+		return err
+	}
+
+	ok := rttStats(raw, topic)
+	if !ok {
+		return fmt.Errorf("Failed to calculate broker rtt statistics")
```

I think it is better to return this error from rttStats instead of a bool so that the actual error is not lost and converted to this generic one.